### ISSUE TITLE
Dynamic destinations with mutiple binders:

### ISSUE THREAD

**Issue by ennala** [spring-cloud/spring-cloud-stream-binder-kafka#913]
Sorry for posting it here.. couldnt get much help from SO. Please pardon me for that.

I need to consume from different topics on one kafka cluster and forward the message to same topic on another kafka cluster/clusters depending on the business logic.

I found an example for BinderAwareChannelResolver for solving destinations dynamically in the documentation .... but not able to understand how to make it work with multiple binders. Appreciate any inputs on how to configure BinderAwareChannelResolver with multiple binders.

Here is more detailed explanation of what I am trying to achieve.

Consume from topic1,topic2,topic3 etc from kafka cluster1.

Forward the message to same topic to kafka cluster 2 if it meets criteria 1.

Forward the message to same topic to kafka cluster 2 if it meets criteria 2.

**Comment by sobychacko** [comment]
@ennala Sorry for not responding to your question on SO. We might have missed it. Do you have a link to that? 

You can use the newly introduced `StreamBridge` instead of `BinderAwareChannelResolver` which is deprecated. You need to use this in addition to the multi binder configuration. 

Here are some templates and pseudo-code:

Assuming that, you have two input topics and two output topics. You can use Tuple[N] classes from Project Reactor for doing this.

```
@Autowired StreamBridge streamBridge;

@Bean
public  Consumer<Tuple2<String, String>> multiInWithDynamicDestination() {
	return tuple -> {
		String first = tuple.getT1();
		String second = tuple.getT2();
              if (first == "some conditions met") {
                streamBrige.send("foo-out-0", first); // foo-out-0 is the binding name.
             }
            else if (second == "some condition...") {
                streamBrige.send("bar-out-0", second); // bar-out-0 is the binding name.
            }				
	};
}
```

Configuration:
```
spring.cloud.stream.source=foo-out-0;bar-out-0
spring:
  cloud:
    stream:
      bindings:
        multiInWithDynamicDestination-in-0:
          destination: dest-1
          binder: kafka1
        multiInWithDynamicDestination-in-1:
          destination: dest-2
          binder: kafka1
        foo-out-0:
          destination: dest-1
          binder: kafka2
       bar-out-0:
          destination: dest-2
          binder: kafka2
      function:
        definition: multiInWithDynamicDestination
      binders:
        kafka1:
          type: kafka
          environment:
            spring:
              cloud:
                stream:
                  kafka:
                    binder:
                      brokers: ${kafkaBroker1}
                      zkNodes: ${zk1}
        kafka2:
          type: kafka
          environment:
            spring:
              cloud:
                stream:
                  kafka:
                    binder:
                      brokers: ${kafkaBroker2}
                      zkNodes: ${zk2}
```

See [this test](https://github.com/spring-cloud/spring-cloud-stream/blob/master/spring-cloud-stream/src/test/java/org/springframework/cloud/stream/function/StreamBridgeTests.java) for more usage examples for `StreamBridge`.

See [this sample](https://github.com/spring-cloud/spring-cloud-stream-samples/blob/master/multi-binder-samples/multi-binder-two-kafka-clusters/src/main/resources/application.yml) for configuring multi binders.

**Comment by ennala** [comment]
@sobychacko thanks for your detailed explanation.

Here is the link [SO](https://stackoverflow.com/questions/62054845/dynamic-destinations-on-mutiple-binders). I will update the thread with the solution from here.

We have 20 different topics and need to produce to 4 different kafka clusters. Is there a way we can dynamically resolve the destination at run time instead of config while producing when using StreamBridge ?

Thanks.

**Comment by sobychacko** [comment]
If you don't provide your binding intention through `spring.cloud.stream.source`, then SCSt will assume that the first argument for `streamBridge.send(...)` is the actual destination name (Kafka topic name in your case), and bind with default properties. The problem with this approach is that it will cause issues in the multi binder scenario. Since we are not providing an explicit binding, rather dynamic one at runtime, it will not know which binder (thus Kafka broker) to use. It will use the default one accessible through the main context. I will see if there is a way to get around this from the framework side, but if it can't be done I am afraid you have to statically provide the binding names.

**Comment by sobychacko** [comment]
`StreamBridge` supports all operations supported by `BinderAwareChannelResolver`, but I guess even for `BinderAwareChannelResolver` there was this problem I mentioned about multi binders. How did you configure the multibinder configuration for that?

**Comment by ennala** [comment]
@sobychacko yeah I was not able to make it work. Thanks

**Comment by sobychacko** [comment]
@ennala I updated [this comment ](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/913#issuecomment-635445489) with some changes and new details. Please check it. 

Your use case is that you have 20 output topics that need to produce into 4 different Kafka clusters. 
The information about the clusters must be provided through configuration as in the multi binder examples. Then you can also configure the output bindings as shown in the example. I understand that in your case, this is going to be 20 literals separated by a semicolon as the value for `spring.cloud.stream.source`.  For example - `spring.cloud.stream.source: order;customer;account;person....`

The other option is to provide an overloaded `send` method on `StreamBridge` that takes a third parameter that indicates the target broker/binder in a map or similar data structure. 

Bottom line is that you have to let the application know about the target brokers either through configuration (already supported as shown ^^) or programmatically. In my opinion, the configuration approach is better and it gives you the flexibility of changing the output bindings at deployment time.  

I am not sure how wide of a use case is this, in order to warrant an API level change in `StreamBridge` to accommodate this. Therefore, my inclination is to go with the configuration approach.  /cc @olegz.

Please let us know what you think. Thank you!

**Comment by sobychacko** [comment]
@ennala Closing this issue now. Feel free to re-open if the suggested solution is not sufficient or create a new issue in the core Spring Cloud Stream [repository](https://github.com/spring-cloud/spring-cloud-stream).
